In [1]:
import pandas as pd
import numpy as np

In [2]:
fdata = pd.read_csv('flight_data.csv')
fdata.head()

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0  2013      1    1     517.0             515        2.0     830.0   
1  2013      1    1     533.0             529        4.0     850.0   
2  2013      1    1     542.0             540        2.0     923.0   
3  2013      1    1     544.0             545       -1.0    1004.0   
4  2013      1    1     554.0             600       -6.0     812.0   

   sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0             819       11.0      UA    1545  N14228    EWR  IAH     227.0   
1             830       20.0      UA    1714  N24211    LGA  IAH     227.0   
2             850       33.0      AA    1141  N619AA    JFK  MIA     160.0   
3            1022      -18.0      B6     725  N804JB    JFK  BQN     183.0   
4             837      -25.0      DL     461  N668DN    LGA  ATL     116.0   

   distance  hour  minute         time_hour  
0      1400     5      15  01-01-2013 05:00  
1      1416     5      29  01-01-2013 05:00  
2      1089     5      40  01-01-2013 05:00  
3      1576     5      45  01-01-2013 05:00  
4       762     6       0  01-01-2013 06:00

In [3]:
fdata.dtypes

year                int64
month               int64
day                 int64
dep_time          float64
sched_dep_time      int64
dep_delay         float64
arr_time          float64
sched_arr_time      int64
arr_delay         float64
carrier            object
flight              int64
tailnum            object
origin             object
dest               object
air_time          float64
distance            int64
hour                int64
minute              int64
time_hour          object
dtype: object

# Check Null Values in Dataset  : 



Is there any null values in Dataset ?

In [4]:
fdata.isnull().values.any()

True

Total Number of Null Values in columns:

In [5]:
fdata.isnull().sum()

year                 0
month                0
day                  0
dep_time          8255
sched_dep_time       0
dep_delay         8255
arr_time          8713
sched_arr_time       0
arr_delay         9430
carrier              0
flight               0
tailnum           2512
origin               0
dest                 0
air_time          9430
distance             0
hour                 0
minute               0
time_hour            0
dtype: int64

In [6]:
fdata['arr_time'].isnull().sum() - fdata['dep_time'].isnull().sum()

458

* There are 458 Flights for which arrival time is NOT available but departure time is available. These are valid number of flights where dep_time data is missing. -- Keep these rows


In [7]:
len(fdata[fdata['arr_time'].isnull() & fdata['dep_time'].isnull()])

8255

* There are 8255 rows in which both arrival & departure time is not available. We can drop above rows from dataset! 

In [8]:
fdata.shape

(336776, 19)

In [9]:
fdata.dropna(subset=['dep_time', 'arr_time'], how='all',inplace=True)

In [10]:
fdata.shape


(328521, 19)

#  Exploration ideas : 
# 1. <U> Departure Delays 

In [11]:
fdata["dep_time"]= pd.to_numeric(fdata['dep_time']).fillna(0).astype(np.int64).apply(lambda x: '{0:0>4}'.format(x))
fdata["sched_dep_time"]= pd.to_numeric(fdata['sched_dep_time']).fillna(0).astype(np.int64).apply(lambda x: '{0:0>4}'.format(x))
fdata["arr_time"]= pd.to_numeric(fdata['arr_time']).fillna(0).astype(np.int64).apply(lambda x: '{0:0>4}'.format(x))
fdata["sched_arr_time"]= pd.to_numeric(fdata['sched_arr_time']).fillna(0).astype(np.int64).apply(lambda x: '{0:0>4}'.format(x))

In [12]:
fdata.head()

year  month  day dep_time sched_dep_time  dep_delay arr_time  \
0  2013      1    1     0517           0515        2.0     0830   
1  2013      1    1     0533           0529        4.0     0850   
2  2013      1    1     0542           0540        2.0     0923   
3  2013      1    1     0544           0545       -1.0     1004   
4  2013      1    1     0554           0600       -6.0     0812   

  sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0           0819       11.0      UA    1545  N14228    EWR  IAH     227.0   
1           0830       20.0      UA    1714  N24211    LGA  IAH     227.0   
2           0850       33.0      AA    1141  N619AA    JFK  MIA     160.0   
3           1022      -18.0      B6     725  N804JB    JFK  BQN     183.0   
4           0837      -25.0      DL     461  N668DN    LGA  ATL     116.0   

   distance  hour  minute         time_hour  
0      1400     5      15  01-01-2013 05:00  
1      1416     5      29  01-01-2013 05:00  
2      1089     5      40  01-01-2013 05:00  
3      1576     5      45  01-01-2013 05:00  
4       762     6       0  01-01-2013 06:00

In [13]:
fdata["dep_time"]= pd.to_timedelta(fdata["dep_time"].str[:2]+":"+fdata["dep_time"].str[-2:]+":00")
fdata["sched_dep_time"]= pd.to_timedelta(fdata["sched_dep_time"].str[:2]+":"+fdata["sched_dep_time"].str[-2:]+":00")
fdata["arr_time"]= pd.to_timedelta(fdata["arr_time"].str[:2]+":"+fdata["arr_time"].str[-2:]+":00")
fdata["sched_arr_time"]= pd.to_timedelta(fdata["sched_arr_time"].str[:2]+":"+fdata["sched_arr_time"].str[-2:]+":00")

In [14]:
fdata.head()

year  month  day dep_time sched_dep_time  dep_delay arr_time  \
0  2013      1    1 05:17:00       05:15:00        2.0 08:30:00   
1  2013      1    1 05:33:00       05:29:00        4.0 08:50:00   
2  2013      1    1 05:42:00       05:40:00        2.0 09:23:00   
3  2013      1    1 05:44:00       05:45:00       -1.0 10:04:00   
4  2013      1    1 05:54:00       06:00:00       -6.0 08:12:00   

  sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0       08:19:00       11.0      UA    1545  N14228    EWR  IAH     227.0   
1       08:30:00       20.0      UA    1714  N24211    LGA  IAH     227.0   
2       08:50:00       33.0      AA    1141  N619AA    JFK  MIA     160.0   
3       10:22:00      -18.0      B6     725  N804JB    JFK  BQN     183.0   
4       08:37:00      -25.0      DL     461  N668DN    LGA  ATL     116.0   

   distance  hour  minute         time_hour  
0      1400     5      15  01-01-2013 05:00  
1      1416     5      29  01-01-2013 05:00  
2      1089     5      40  01-01-2013 05:00  
3      1576     5      45  01-01-2013 05:00  
4       762     6       0  01-01-2013 06:00

In [15]:
fdata['fdate'] = fdata['year'].astype(str) + "/" + fdata['month'].astype(str) + "/" + fdata['day'].astype(str)


128432

In [18]:
len(fdata[fdata.dep_delay>0])

In [27]:
delay_Flights = fdata[fdata.dep_delay>0]


In [26]:
delay_count= pd.DataFrame(delay_Flights.groupby('fdate').count())


In [24]:
delay_count.head()

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
fdate                                                                        
2013/1/1    352    352  352       352             352        352       352   
2013/1/10   191    191  191       191             191        191       191   
2013/1/11   237    237  237       237             237        237       237   
2013/1/12   173    173  173       173             173        173       173   
2013/1/13   363    363  363       363             363        363       363   

           sched_arr_time  arr_delay  carrier  flight  tailnum  origin  dest  \
fdate                                                                          
2013/1/1              352        347      352     352      352     352   352   
2013/1/10             191        191      191     191      191     191   191   
2013/1/11             237        237      237     237      237     237   237   
2013/1/12             173        172      173     173      173     173   173   
2013/1/13             363        358      363     363      363     363   363   

           air_time  distance  hour  minute  time_hour  
fdate                                                   
2013/1/1        347       352   352     352        352  
2013/1/10       191       191   191     191        191  
2013/1/11       237       237   237     237        237  
2013/1/12       172       173   173     173        173  
2013/1/13       358       363   363     363        363

In [ ]:
mydate = pd.DataFrame(delay_count['dep_delay'])
mydate.loc[mydate['dep_delay'].idxmax()]

* <b>Conclusion 1:</b> Maximum  number of flight delays [674] on <b>2013/12/23</b>

In [29]:
fdata.dtypes

year                        int64
month                       int64
day                         int64
dep_time          timedelta64[ns]
sched_dep_time    timedelta64[ns]
dep_delay                 float64
arr_time          timedelta64[ns]
sched_arr_time    timedelta64[ns]
arr_delay                 float64
carrier                    object
flight                      int64
tailnum                    object
origin                     object
dest                       object
air_time                  float64
distance                    int64
hour                        int64
minute                      int64
time_hour                  object
fdate                      object
dtype: object

In [ ]:
fdata["dep_delay"]= pd.to_numeric(fdata['dep_delay']).fillna(0).astype(np.int64)


In [ ]:
delay_time_max = fdata['dep_delay'].max()
delay_time_max

In [ ]:

fdata[fdata['dep_delay']==fdata['dep_delay'].max()]


In [ ]:
fdata['fdate'].loc[fdata['dep_delay'].idxmax()]

* <b>Conclusion 2:</b> Maximum Time [1301 minutes] of flight delay happend on <b>2013/1/9</b>

# 2. Best airport in terms of time departure %

# 3. Aircraft speed analysis


# 4. On time arrival % analysis

# 5. Maximum number of flights headed to some perticular destination